### ISSUE TITLE
Multiple `preConditions` blocks or a group within

### ISSUE THREAD

**Issue by Vampire** [liquibase/liquibase#3116]
## Environment

**Liquibase Version**: 3.6.3

**Liquibase Integration & Version**: As library in own updater tool

**Liquibase Extension(s) & Version**: n/a

**Database Vendor & Version**: n/a

**Operating System Type & Version**: n/a

**Infrastructure Type/Provider**: n/a

From time to time I need to have preconditions result in different outcomes.
It would be nice if you for example could either have multiple `preConditions` blocks, or some way to group within.
This should then enable things like

```xml
<preConditions onFail="MARK_RAN">
   <not>
      <changeSetExecuted id="a" author="b"/>
   </not>
</preConditions>
<preConditions>
   <tableExists tableName="A"/>
   <not>
      <columnExists tableName="A" columnName="B"/>
   </not>
</preConditions>
<addColumn tableName="A">
   <column name="B" type="VARCHAR(255)"/>
</addColumn>
```
or
```xml
<preConditions>
   <group onFail="MARK_RAN">
      <not>
         <changeSetExecuted id="a" author="b"/>
      </not>
   </group>
   <group>
      <tableExists tableName="A"/>
      <not>
         <columnExists tableName="A" columnName="B"/>
      </not>
   </group>
</preConditions>
<addColumn tableName="A">
   <column name="B" type="VARCHAR(255)"/>
</addColumn>
```
or similar.

**Comment by b-gyula** [comment]
You can have multiple `preConditions` on the `databaseChangeLog` level in ALL formats except XML

**Comment by filipelautert** [comment]
We're currently triaging our backlog to ensure we can focus on the issues that matter most. With a large number of open issues, it becomes difficult to give each one the attention it deserves.

This issue has been closed due to inactivity — there have been no updates for over 30 days. If this is still relevant to you, please feel free to reopen it or create a new issue with updated context. We appreciate your contribution!

**Comment by Vampire** [comment]
@filipelautert I can't reopen an issue you closed, only tickets I closed myself. 

There were no updates because all was said and now it would be nice if you provide the feature. I don't see any need for additional context from my side.

**Comment by filipelautert** [comment]
Thanks for the feature request and for the continued interest from @Vampire and @b-gyula.

After investigating, this is a significant feature enhancement that would require changes across multiple layers (XSD schema, precondition evaluation model, all changelog parsers). The current architecture treats preconditions as a single block per changeSet with uniform onFail/onError behavior by design.

The existing workaround for achieving different failure behaviors is to split the logic into separate changeSets, each with its own preConditions block and onFail setting. While less convenient than the proposed grouped approach, it achieves the same result:

```xml
<changeSet id="check-a" author="x">
    <preConditions onFail="MARK_RAN">
        <not><changeSetExecuted id="a" author="b"/></not>
    </preConditions>
    <comment>Guard: skip if changeset a was already executed</comment>
</changeSet>

<changeSet id="add-col-B" author="x">
    <preConditions>
        <tableExists tableName="A"/>
        <not><columnExists tableName="A" columnName="B"/></not>
    </preConditions>
    <addColumn tableName="A">
        <column name="B" type="VARCHAR(255)"/>
    </addColumn>
</changeSet>
```

Closing as not planned since the scope of this change is large relative to the availability of a reasonable workaround. If there is strong community demand, we can revisit this as a formal design proposal.

**Comment by Vampire** [comment]
@filipelautert mind elaborating on that work-around?
How is `add-col-B` now skipped if changeset `a` was already executed?

Maybe I'm tired, but I see two independent changesets,
And if `a` was not executed yet, `check-a` is recorded as executed wihtout doing anything.
While `add-col-B` will always be executed and expect column `B` to be absent.

Just in case it was not clear, changeset `a` did already add column `B` in the correct shape.
So if changeset `a` was executed, the `addColumn` should be `MARK_RAN`.
But if `a` was not executed, the `B` must not exist or we don't know where it comes from and thus it should fail if found in that case.

I really don't see how your suggested work-around achieves that.

**Comment by rberezen** [comment]
@Vampire you're right — the suggested workaround with separate changeSets doesn't cover your use case, since the second changeSet loses the conditional context of the first. Apologies for the confusion.

To summarize the requirement: you need different `onFail` outcomes from preconditions within the same changeSet — `MARK_RAN` if changeset `a` was already executed, but `HALT` if table/column state is unexpected. This can't be expressed with the current single `onFail` per `preConditions` block.

We're keeping this open as a valid enhancement request.
